ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureStorageCompactor.java

```diff
@@ -257,17 +309,61 @@ int purgeBlobs(List<CloudBlobMetadata> blobMetadataList) throws CloudStorageExce
   }
 
   /**
-   * Returns the dead blob in the specified partition with the earliest expiration or deletion time.
+   * Get the current compaction progress for a partition.
    * @param partitionPath the partition to check.
-   * @param fieldName the field name to use (expiration or deletion time).
-   * @return the {@link CloudBlobMetadata} for the dead blob, or NULL if none was found.
-   * @throws CloudStorageException
+   * @return a {@link Map} containing the progress time for compaction based on deletion and expiration time.
+   * @throws BlobStorageException if the checkpoint blob
+   */
+  Map<String, Long> getCompactionProgress(String partitionPath) throws BlobStorageException {
+    ByteArrayOutputStream baos = new ByteArrayOutputStream(64);
+    boolean hasCheckpoint = azureBlobDataAccessor.downloadFile(CHECKPOINT_CONTAINER, partitionPath, baos, false);
+    if (!hasCheckpoint) {
+      return emptyCheckpoints;
+    }
+    try {
+      // Payload format: {"expirationTime" : 12345, "deletionTime" : 67890}
+      ObjectNode jsonNode = (ObjectNode) objectMapper.readTree(baos.toString());
+      Map<String, Long> checkpoints = new HashMap<>();
+      for (String fieldName : new String[]{CloudBlobMetadata.FIELD_DELETION_TIME,
+          CloudBlobMetadata.FIELD_EXPIRATION_TIME}) {
+        checkpoints.put(fieldName, jsonNode.has(fieldName) ? jsonNode.get(fieldName).longValue() : DEFAULT_TIME);
+      }
+      return checkpoints;
+    } catch (IOException e) {
+      logger.error("Could not retrieve compaction progress for {}", partitionPath, e);
+      azureMetrics.compactionProgressReadErrorCount.inc();
+      return emptyCheckpoints;
+    }
+  }
+
+  /**
+   * Update the compaction progress for a partition.
+   * @param partitionPath the partition to update.
+   * @param fieldName the compaction field (deletion or expiration time).
+   * @param checkpointTime the updated progress time.
    */
-  public CloudBlobMetadata getOldestDeadlob(String partitionPath, String fieldName) throws CloudStorageException {
-    // TODO: once we have Cosmos compaction table, can query that.
-    List<CloudBlobMetadata> deadBlobs =
-        requestAgent.doWithRetries(() -> getDeadBlobs(partitionPath, fieldName, 1, System.currentTimeMillis(), 1),
-            "GetDeadBlobs", partitionPath);
-    return deadBlobs.isEmpty() ? null : deadBlobs.get(0);
+  boolean updateCompactionProgress(String partitionPath, String fieldName, long checkpointTime) {
+    try {
+      // load existing progress checkpoint.
+      Map<String, Long> checkpoints = getCompactionProgress(partitionPath);
```

Comment:
We should cache this information instead of doing a get always. Only when we dont see a cache for this partition, should we do `getCompactionInProgress`

Reply:
I'm hesitant to cache it since the partition could get reassigned back and forth between runs and be out of date.   I'd probably want to read it anyway to make sure.